### ISSUE TITLE
Hook to optionally skip specific failures

### ISSUE THREAD

**Issue by Stranger6667** [schemathesis/schemathesis#1792]
Some failures are not the API's fault and not worth enumerating one by one - an upstream returning 503, a body reporting that a dependency timed out. 

Per-check config, scoped by `[[operations]]`, selects by rule and operation - not by response content.

### API

A predicate in the existing `filter_*` family. Return False to drop the failure:

```python
@schemathesis.hook
def filter_failure(ctx, failure, case, response):
    return response.status_code != 503
```

Runs once per failure, after the checks for a response.

A dropped failure does not fail the run, does not spend `--max-failures`, does not enter deduplication, and does not create a crash file. It is counted and reported alongside baseline matches rather than getting a status of its own.

**Out of scope**: network errors and after_run check failures, which have no case/response pair.

**Comment by Stranger6667** [comment]
It could be a hook like this:

```python
import schemathesis


@schemathesis.hook
def something_something_bikeshed(context, error, case, response):
    # Check whether this should be skipped
    context.skip()
```

**Comment by mxsrc** [comment]
I'd appreciate this feature. If there was a convenient way of skipping specific tests from the CLI that would be great as well, e.g. by specifying a test case ID. Likely they will change with a different seed, but maybe there's an option for a stable identifier?

**Comment by vemolista** [comment]
This would be great.

My use case is that I'm testing a GraphQL API. It is a service where the team prioritized speed, so now we have some tech debt to pay down. So first, I'm trying to figure out what _kinds_ of errors we run into.

For that, it would be handy to be able to skip "known" errors, so schemathesis can continue without cluttering the output with issues that I know about.

The workaround for now is that we use `--continue-on-failure` and then go through all of the errors we got.

**Comment by patpatpat123** [comment]
# Proposal: conditionally ignore known failures (`#1792`)

Hi everyone! I'd like to work on #1792. Before starting the implementation, here is the proposed behavior and planned code diff so we can confirm the direction first.

## What I'm proposing

Introduce the hook described in the issue:

```python
import schemathesis


@schemathesis.hook
def ignore_failure_if(context, error, case, response):
    if response.status_code == 503:
        context.ignore(error)
```

The hook will run once for each individual failure produced by a response check.

An ignored failure will:

- Not stop Hypothesis from generating more cases.
- Not make the scenario, phase, or command fail.
- Not increment `--max-failures`.
- Not enter stateful failure deduplication.
- Not create a crash file.
- Remain visible as an ignored check in CLI output and lifecycle hooks.

If one response produces both ignored and non-ignored failures, the request will still fail, but only the non-ignored failures will appear in the failure report.

## Proposed semantics

| Situation | Result |
|---|---|
| All checks pass | Success |
| One or more failures are ignored | Success with an ignored-failure annotation |
| Ignored and real failures occur together | Failure; only real failures are reported |
| The ignore hook raises an exception | Execution error |
| Network request fails | Existing network-error behavior |
| An `after_run` check fails | Existing run-level failure behavior |

Network errors and `after_run` failures are intentionally outside the initial scope because they do not have the `case` and `response` pair required by the proposed hook.

## Planned implementation diff

The snippets below are illustrative rather than a literal final patch, but they show where the behavior will live.

### 1. Add the hook API

Files:

- `src/schemathesis/hook_specs.py`
- `src/schemathesis/hooks.py`

```diff
 @all_scopes
 def after_validate(context, case, response, results):
     ...

+@all_scopes
+def ignore_failure_if(context, error, case, response):
+    """Conditionally ignore a response-check failure."""
+    ...
```

`HookContext` will keep the ignore decision local to the current failure:

```diff
 class HookContext:
     operation: APIOperation | None
+    _ignored_failures: set[Failure]

+    def ignore(self, error: Failure) -> None:
+        self._ignored_failures.add(error)
+
+    def is_ignored(self, error: Failure) -> bool:
+        return error in self._ignored_failures
```

The dispatcher will call matching hooks in the existing order:

```text
global hooks -> schema hooks -> test-local hooks
```

Hook filters such as `.apply_to(...)` and `.skip_for(...)` will work normally. Any matching hook may ignore the failure; later hooks cannot "unignore" it.

### 2. Make the decision at the central check boundary

File:

- `src/schemathesis/checks.py`

This is the important part of the implementation. Check exceptions are already flattened into individual `Failure` objects here.

```diff
 for failure in _failures_from_exception(name, case.operation.label, exc):
-    on_failure(name, collected, failure)
+    if dispatch_ignore_failure_if(
+        case=case,
+        response=response,
+        error=failure,
+    ):
+        if on_ignored is not None:
+            on_ignored(name, failure)
+    else:
+        on_failure(name, collected, failure)
```

Doing this before `on_failure` means ignored failures never reach:

- `FailureGroup`
- Hypothesis failure handling
- Engine failure counters
- Unique-input failure caches
- Stateful seen-failure sets
- Normal failure statistics

Each member of a `FailureGroup` is evaluated independently, which allows one validation error to be ignored while another remains actionable.

### 3. Record ignored checks without changing scenario outcomes

Files:

- `src/schemathesis/engine/__init__.py`
- `src/schemathesis/engine/recorder.py`
- `src/schemathesis/engine/statistic.py`

Add an ignored status for individual check results:

```diff
 class Status(str, Enum):
     SUCCESS = "success"
     FAILURE = "failure"
+    IGNORED = "ignored"
     ERROR = "error"
```

`Status.IGNORED` will only be used for check results and recorder nodes. An ignored-only scenario will still finish with `Status.SUCCESS`.

```diff
 class ScenarioRecorder:
+    def record_check_ignored(
+        self,
+        *,
+        name,
+        case_id,
+        code_sample,
+        failure,
+    ):
+        self.checks.setdefault(case_id, []).append(
+            CheckNode(
+                name=name,
+                status=Status.IGNORED,
+                failure_info=CheckFailureInfo(
+                    code_sample=code_sample,
+                    failure=failure,
+                ),
+            )
+        )
```

Failure aggregation will explicitly select `Status.FAILURE`. An ignored node may retain failure details for observability without accidentally appearing in normal failure reports.

Statistics will additionally track how many executed cases contained ignored failures.

### 4. Connect all response-validation paths

Files:

- `src/schemathesis/generation/case.py`
- `src/schemathesis/engine/_validate.py`
- `src/schemathesis/engine/run/stateful/_executor.py`
- `src/schemathesis/pytest/reporting.py`

Each validation path will provide an `on_ignored` callback:

```diff
 failures = run_checks(
     case=case,
     response=response,
     on_failure=on_failure,
+    on_ignored=on_ignored,
     on_success=on_success,
 )
```

For the public Python API, `after_validate` will receive:

```python
CheckResult(
    name="response_schema_conformance",
    status=Status.IGNORED,
    failure=error,
)
```

For stateful testing, ignored failures will be recorded but will not:

- Increment the stateful failure counter.
- Enter `seen_in_suite` or `seen_in_run`.
- Stop or restart the state machine.

The same central behavior will apply to examples, coverage, fuzzing, `st fuzz`, pytest, and direct `Case.call_and_validate()` usage.

### 5. Add friendly CLI annotations

Files:

- `src/schemathesis/cli/output.py`
- `src/schemathesis/cli/commands/run/handlers/output.py`
- `src/schemathesis/cli/commands/fuzz/handlers/output.py`

Ignored-only cases remain part of the passed total, with a separate annotation:

```text
✅ 25 passed  ⚪ 3 with ignored failures
```

The test-case summary will follow the same model:

```text
Test cases:
  25 generated, 25 passed, 3 with ignored failures
```

When a run has no real failures:

```text
No unignored issues found in 2.31s (3 cases had ignored failures)
```

The command will exit successfully.

Ignored failures will not be printed in the normal `FAILURES` section because that would make known failures indistinguishable from actionable ones.

### 6. Keep existing report formats safe

Files with recorder/report consumption will be audited so they only treat nodes with `Status.FAILURE` as failures.

Expected behavior:

| Output | Ignored-only case |
|---|---|
| CLI | Passed, with ignored annotation |
| Pytest | Passed |
| JUnit | Passed |
| Allure | Passed |
| VCR | Check stored with `IGNORED` status |
| NDJSON | Check stored with `ignored` status |
| Crash files | Not created |
| Pytest-xdist | Ignored node survives serialization |

This initial change does not introduce a new JUnit or Allure outcome. Those formats retain their existing pass/fail model.

## Tests I plan to add

- Hook registration at global, schema, and test scope.
- Hook filtering with `.apply_to(...)` and `.skip_for(...)`.
- Ignoring a built-in failure.
- Ignoring a custom-check `AssertionError`.
- Independently filtering failures flattened from a `FailureGroup`.
- A response containing both ignored and real failures.
- Hook exceptions remaining execution errors.
- `after_validate` receiving `Status.IGNORED`.
- Ignored failures not affecting `--max-failures`.
- Ignored failures not stopping examples, coverage, fuzzing, `st fuzz`, or stateful testing.
- Ignored failures not entering stateful seen-failure sets or unique-input failure caches.
- CLI snapshots for ignored-only and mixed outcomes.
- JUnit and Allure treating ignored-only scenarios as passed.
- VCR, NDJSON, and pytest-xdist preserving the ignored check status.
- Crash reporting excluding ignored failures.

After the focused tests pass, I'll run the complete project test suite.

## Documentation

I'll update the hooks reference and extension guide with:

- The complete hook signature.
- A practical example for ignoring known server failures.
- Hook ordering and filtering behavior.
- The distinction between ignored and skipped cases.
- Mixed ignored/non-ignored behavior.
- The response-check-only scope.
- The fact that ignored failures are still observable through `after_validate`.

## Feedback welcome

The implementation will follow the behavior above unless reviewers prefer a different direction. In particular, feedback would be very helpful on:

- The `context.ignore(error)` API.
- The proposed CLI wording.
- Treating ignored-only cases as passed in JUnit and Allure.
- Keeping network and run-level failures outside the first version.

Thanks for taking a look!
